kombu/transport/cherami.py

```diff
@@ -0,0 +1,151 @@
+
+        # event_loop
+        self.hub = kwargs.get('hub') or get_event_loop()
+
+        # cherami-client instance kwargs
+        self.kwargs = self.connection.kwargs
+
+        # delivery_tag(Kombu.Transport.QoS) -> delivery_token(cherami)
+        self.delivery_map = dict()
+
+    def basic_consume(self, queue, no_ack, *args, **kwargs):
+        # set the prefetch_limit specified by the celery configuration
+        self.prefetch_limit = self.qos.prefetch_count
+
+        if self.hub:
+            self.hub.call_soon(self._get, queue)
+        # no_ack is True because cherami is in charge of message ack
+        return super(Channel, self).basic_consume(
+            queue, True, *args, **kwargs
+        )
+
+    def _put(self, queue, message, **kwargs):
+        # the first argument is the message id, cherami doesn't care about this, just passes along
+        self.publisher.publish(str(0), dumps(message))
+
+    def _get(self, queue, callback=None):
+        if self.prefetched < self.prefetch_limit:
+            try:
+                results = self.consumer.receive(num_msgs=self.default_message_batch_size)
+                self._on_message_ready(queue, results)
+            except Exception as e:
+                logger.info('Failed to receive messages: {0}'.format(e))
+        else:
+            self.hub.call_soon(self._get, queue)
+
+    def _on_message_ready(self, queue, results):
+        for res in results:
+            delivery_token = res[0]
+            message = res[1]
+            try:
+                self._handle_message(queue, message.payload.data, delivery_token)
+            except Exception as e:
+                self.consumer.nack(delivery_token)
+                logger.info('Failed to process a message:  {0}'.format(e))
+        # done processing messages, consume again
+        self.hub.call_soon(self._get, queue)
+
+    def _handle_message(self, queue, data, delivery_token):
+        message = loads(data)
+
+        # saves the mapping for ack
+        delivery_tag = message['properties']['delivery_tag']
+        self.delivery_map[delivery_tag] = delivery_token
+
+        self.prefetched += 1
+        self.connection._deliver(message, queue)
+
+    def basic_ack(self, delivery_tag, multiple=False):
+        # get the delivery_token for cherami ack
+        delivery_token = self.delivery_map[delivery_tag]
+        self.consumer.ack(delivery_token)
+
+        # removes the mapping
+        self.delivery_map.pop(delivery_tag, None)
+        self.prefetched -= 1
+
+    def close(self):
+        super(Channel, self).close()
+        if self._consumer:
+            self._consumer.close()
+        if self._publisher:
+            self._publisher.close()
+
+    @property
+    def publisher(self):
+        if self._publisher is None:
+            if isinstance(self.kwargs['cherami_publisher'], Publisher):
+                self._publisher = self.kwargs['cherami_publisher']
+                self._publisher.open()
```

Comment:
as we discussed, it's better to have the open/close logic to be the same place where the object is created.